File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/base/UaReadTransaction.java
package org.opcfoundation.webserver.service.transactions.base;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.DataValue;
import org.eclipse.milo.opcua.stack.core.types.builtin.StatusCode;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.enumerated.TimestampsToReturn;
import org.eclipse.milo.opcua.stack.core.types.structured.ReadValueId;
import org.opcfoundation.webapi.service.types.ReadContext;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class UaReadTransaction extends UaTransaction2<ReadValueId, DataValue> {
    protected final TimestampsToReturn timestampsToReturn;
    protected List<DataValue> results;

    public UaReadTransaction(
            ReadContext readContext,
            List<Integer> handleIds)
    {
        super(readContext,handleIds);
        this.timestampsToReturn = readContext.getTimestampsToReturn();
        results = new ArrayList<>();
    }

    public List<ReadValueId> getItems()
    {
        return ((ReadContext)serviceContext).getNodesToRead();
    }

    public List<DataValue> getResults()
    {
        return results;
    }

    public CompletableFuture<Void> execute()
    {
        for (Integer item: getHandleIds())
        {
            results.add(new DataValue(Variant.NULL_VALUE, StatusCode.of(StatusCodes.Bad_NodeIdUnknown), null));
        }
        return CompletableFuture.completedFuture(null);
    }

    protected Void buildErrorResponse(Throwable exception)
    {
        StatusCode errorCode = StatusCode.of(StatusCodes.Bad_UnexpectedError);

        if (exception instanceof UaRuntimeException)
        {
            errorCode = ((UaRuntimeException)exception).getStatusCode();
        }

        for (Integer item: getHandleIds())
        {
            results.add(new DataValue(Variant.NULL_VALUE, errorCode, null));
        }

        return null;
    }
}

File: opcua-webservice-node/src/server/service/transactions/base/UaReadTransaction.ts
import { StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaDataValue, UaError, UaReadValueId, UaVariant } from "opcua-webapi-ts";
import { ReadContext } from "../../..";
import { UaTransaction2 } from ".";

export class UaReadTransaction extends UaTransaction2<UaReadValueId, UaDataValue> {
    protected readonly _timestampsToReturn: number;
    protected _results: Array<UaDataValue>;

    constructor(readContext: ReadContext, handleIds: Array<number>) {
        super(readContext, handleIds);
        this._timestampsToReturn = readContext.timestampsToReturn;
        this._results = [];
    }

    get timestampsToReturn(): number {
        return this._timestampsToReturn;
    }

    getItems(): Array<UaReadValueId> {
        return (this.serviceContext as ReadContext).nodesToRead;
    }

    getResults(): Array<UaDataValue> {
        return this._results;
    }

    async execute(): Promise<void> {
        for (const _handleId of this.handleIds) {
            this._results.push(
                new UaDataValue(
                    UaVariant.null(),
                    makeUaStatusCode(StatusCodes.BadNodeIdUnknown),
                ),
            );
        }
    }

    protected buildErrorResults(error: unknown): void {
        let statusCode = makeUaStatusCode(StatusCodes.BadUnexpectedError);
        if (error instanceof UaError) {
            statusCode = error.statusCode;
        } 

        this._results = [];        
        for (const _item of this.getRequestedItems()) {
            this._results.push(new UaDataValue(UaVariant.null(), statusCode));
        }
    }
}